src/sign_tx.c

```diff
@@ -0,0 +1,449 @@
+        }
+        // we require the input data to be empty because we're signing the whole
+        // bytes we get from the wallet (in sighash_all, inputs must have no data)
+        if (U2BE(ctx->buffer, 33) > 0) {
+            THROW(TX_STATE_ERR);
+        }
+        // we ignore it
+        ctx->remaining_inputs--;
+        ctx->buffer_len -= 35;
+        ctx->elem_type = ELEM_INPUT;
+        os_memmove(ctx->buffer, ctx->buffer + 35, ctx->buffer_len);
+        PRINTF("decoded input\n");
+    } else if (ctx->current_output < ctx->outputs_len) {
+        uint8_t *buf = parse_output(ctx->buffer, ctx->buffer_len, &ctx->decoded_output);
+        PRINTF("decoded output\n");
+        ctx->decoded_output.index = ctx->current_output;
+        ctx->elem_type = ELEM_OUTPUT;
+        ctx->buffer_len -= (buf - ctx->buffer);
+        os_memmove(ctx->buffer, buf, ctx->buffer_len);
+        ctx->current_output++;
+    } else {
+        // end of data we should read. Is there something left on the buffer?
+        if (ctx->buffer_len > 0) {
+            THROW(TX_STATE_ERR);
+        }
+        THROW(TX_STATE_FINISHED);
+    }
+
+    switch (ctx->elem_type) {
+        case ELEM_TOKEN_UID:
+            // not displaying token uid now
+            break;
+        case ELEM_INPUT:
+            // not displaying inputs
+            break;
+        case ELEM_OUTPUT:
+            // check if this is the change output
+            if (ctx->has_change_output && ctx->change_output_index == ctx->decoded_output.index) {
+                if (!verify_change_output(ctx->decoded_output, ctx->change_key_index)) {
+                    THROW(TX_STATE_ERR);
+                }
+            } else {
+                // if it's not change output, raise TX_STATE_READY to display output on screen
+                THROW(TX_STATE_READY);
+            }
+            break;
+    }
+}
+
+tx_decoder_state_e decode_next_element() {
+    volatile tx_decoder_state_e result;
+    BEGIN_TRY {
+        TRY {
+            // read until we reach a displayable element or the end of the buffer
+            for (;;) {
+                _decode_next_element();
+            }
+        }
+        CATCH_OTHER(e) {
+            result = e;
+        }
+        FINALLY {
+        }
+    }
+    END_TRY;
+    PRINTF("decode_next_element result %d\n", result);
+    return result;
+}
+
+/*
+ * Parses the change output info and returns its size. The first byte indicates
+ * whether there's change or not (no change output if byte=0x00). There are 2
+ * possible return values:
+ *   . 1: if there's no change output, it's only 1 byte;
+ *   . 6: if there is, it's [change_exists (1 byte) + output_index (1 byte) + key_index (4 bytes)]
+ */
+static uint8_t parse_change_output_info(uint8_t *in, size_t inlen) {
+    uint8_t *buf = in;
+    assert_length(1, inlen);
+    // check output change. If next byte is greater than 0, there's a change output
+    ctx->has_change_output = (*buf > 0 ? true : false);
```

Comment:
It is possible to have multiple "change" outputs. What I mean is that the transaction may be multiple outputs with P2PKH to an address in the wallet.

Reply:
We don't have this in our wallet, that's why it's not being considered here. If we change this behavior in the wallet we can updated it here also.

Reply:
We should make it clear that we are assuming that only one output will be considered change and it will be first with our address. All others will be presented as normal outputs even if they also have our address.

Reply:
Nope, the wallet tells us which one is the change output. That's the one that's not presented. The other outputs, even if they have our address, will be displayed.

Reply:
Now I got it. You're right and I think this is the best behavior. We should just explain it in the design and in the comments.